File: backend/src/main/java/com/sofrecom/backend/entities/Book.java
package com.sofrecom.backend.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
public class Book {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String title;
    private String author;
    @Column(length = 3000)
    private String description;
    private String coverUrl;
    private String publishedDate;
    private String isbn;
    private String category;
    private int pageCount;
    private String language;
    private LocalDate AddedDate;

    @Transient
    private Long ownerId;

    @JsonIgnore
    @ManyToOne
    @JoinColumn(name = "owner_id")
    private User owner;

    @JsonIgnore
    @OneToMany(mappedBy = "book")
    private List<Borrow> borrows;


}

File: app/lib/models/book.dart
class Book {
  final String title;
  final String author;
  final String description;
  final String coverUrl;
  final String publishedDate;
  final String isbn;
  final String category;
  final int pageCount;
  final String language;

  Book({
    required this.title,
    required this.author,
    required this.description,
    required this.coverUrl,
    required this.publishedDate,
    required this.isbn,
    required this.category,
    required this.pageCount,
    required this.language,
  });

  // Conversion d'un JSON en objet Book
  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      title: json['title'] ?? '',
      author: json['author'] ?? '',
      description: json['description'] ?? '',
      coverUrl: json['coverUrl'] ?? '',
      publishedDate: json['publishedDate'] ?? '',
      isbn: json['isbn'] ?? '',
      category: json['category'] ?? '',
      pageCount: json['pageCount'] ?? 0,
      language: json['language'] ?? '',
    );
  }

  // Conversion de l'objet Book en JSON
  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'author': author,
      'description': description,
      'coverUrl': coverUrl,
      'publishedDate': publishedDate,
      'isbn': isbn,
      'category': category,
      'pageCount': pageCount,
      'language': language,
    };
  }

  // Copie d'un objet Book avec modification possible de certains champs
  Book copyWith({
    String? title,
    String? author,
    String? description,
    String? coverUrl,
    String? publishedDate,
    String? isbn,
    String? category,
    int? pageCount,
    String? language,
  }) {
    return Book(
      title: title ?? this.title,
      author: author ?? this.author,
      description: description ?? this.description,
      coverUrl: coverUrl ?? this.coverUrl,
      publishedDate: publishedDate ?? this.publishedDate,
      isbn: isbn ?? this.isbn,
      category: category ?? this.category,
      pageCount: pageCount ?? this.pageCount,
      language: language ?? this.language,
    );
  }
} 